### PR TITLE
gsbabil/fix-restart-command-on-windows

### PR DESCRIPTION
**Pull request by gsbabil**
Dear Maintainers,

The [patches here][0] fixes the `:restart` command on Windows.

Currently `:restart` on Windows requires the user to `:set profiledir` to restart into the desired profile. If `profiledir` is set to `auto` (default behavior), Firefox would still restart, but would show user the profile-selection window. If profile directory is incorrectly set, Firefox won't quit, and show user an error message via `fillcmdline`.

**Note:**

`set profiledir` on Windows must escape profile directory path properly. For example, a correct way to set `profiledir` on Windows would be the following:

```
:set profiledir C:\\Users\\<User-Name>\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\8s21wzbh.Default
```

Hope this helps. Thank you.

[0]: https://github.com/cmcaine/tridactyl/commit/c30327f73e5a539c6d5fedcb4f0ea0a23a761d05

### CODE REVIEW

File: native/native_main.py [hunk 1]
```diff
@@ -16,6 +18,14 @@
 class NoConnectionError(Exception):
```
Comment (bovine3dom) [inline native/native_main.py]: Please bump the version :)

File: src/excmds.ts [hunk 2]
```diff
@@ -400,11 +400,23 @@ export async function updatenative(interactive = true) {
...
 //#background
 export async function restart() {
-    const firefox = (await Native.ffargs()).join(" ")
     const profile = await Native.getProfileDir()
-    // Wait for the lock to disappear, then wait a bit more, then start firefox
-    Native.run(`while readlink ${profile}/lock ; do sleep 1 ; done ; sleep 1 ; ${firefox}`)
-    qall()
+
+    if ((await browser.runtime.getPlatformInfo()).os === "win") {
+        let reply = await Native.winRestartFirefox(profile)
```
Comment (bovine3dom) [inline src/excmds.ts]: Please add an "interactive" version gate here, `nativegate`, to whichever version you bump it to.

File: src/native_background.ts [hunk 3]
```diff
@@ -270,14 +275,34 @@ export async function getenv(variable: string) {
  You'll get both firefox binary (not necessarily an absolute path) and flags */
 export async function ffargs(): Promise<string[]> {
     // Using ' and + rather that ` because we don't want newlines
-    let output = await pyeval(
-        'handleMessage({"cmd": "run", ' +
-            '"command": "ps -p " + str(os.getppid()) + " -oargs="})["content"]',
-    )
-    return output.content.trim().split(" ")
+    if ((await browserBg.runtime.getPlatformInfo()).os === "win") {
+        return []
```
Comment (bovine3dom) [inline src/native_background.ts]: This is somewhat misleading - shouldn't this throw an error?

File: src/native_background.ts [hunk 4]
```diff
@@ -270,14 +275,34 @@ export async function getenv(variable: string) {
...
+    //
+    // : set profiledir C:\\Users\\<User-Name>\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\8s21wzbh.Default
+    if ((await browserBg.runtime.getPlatformInfo()).os === "win") {
+        let win_profiledir = config.get("profiledir")
+        win_profiledir = win_profiledir.trim()
+        if (win_profiledir.length > 0) {
+            return win_profiledir
+        } else {
+            throw new Error("Profile directory not set.")
+        }
```
Comment (bovine3dom) [inline src/native_background.ts]: Somewhere we had a more useful error message telling people go to set their profile and they they could find it on about:support, we should do something similar here.

File: src/native_background.ts [hunk 5]
```diff
@@ -270,14 +275,34 @@ export async function getenv(variable: string) {
...
+    }
 }
 
 export async function getProfileDir() {
+    // Windows users must specify their Firefox profile directory
+    // via 'set profiledir [directory]'. Windows profile directory
+    // path must be properly escaped. For example, a correct way
+    // to set 'profiledir' on Windows is as shown below:
+    //
+    // : set profiledir C:\\Users\\<User-Name>\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\8s21wzbh.Default
```
Comment (bovine3dom) [inline src/native_background.ts]: I'd much rather we escaped it for them. Where does the requirement come from? I think python has `os.path` for dealing with these things.

File: native/native_main.py (hunk 1 above)
Comment (gsbabil) [reply native/native_main.py]: [Done!][0]

[0]: https://github.com/gsbabil/tridactyl/commit/66444af20a548dba07269ed30c39246f4d6acf08#diff-0e71d767312cefb8a12ee5d2753346d1R15

File: src/excmds.ts (hunk 2 above)
Comment (gsbabil) [reply src/excmds.ts]: [Done!][0]

[0]: https://github.com/gsbabil/tridactyl/commit/ef788a393f09c17858e10e42f61eb00f6a7ed96e#diff-9ddd9172e4a54cdd30d56e5f10d5facaR251

File: src/native_background.ts (hunk 3 above)
Comment (gsbabil) [reply src/native_background.ts]: Fixed [here][0]. 

Currently `ffargs()` is unused for Windows specific things. I thought I'd need it to get `:restart` working. But eventually I didn't need it. 

If we need `ffargs()` or alike for other functionalities, I'll submit a different patch. Thanks.

[0]: https://github.com/gsbabil/tridactyl/commit/ef788a393f09c17858e10e42f61eb00f6a7ed96e#diff-9ddd9172e4a54cdd30d56e5f10d5facaR286

File: src/native_background.ts (hunk 4 above)
Comment (gsbabil) [reply src/native_background.ts]: [Done!][0]

[0]: https://github.com/gsbabil/tridactyl/commit/ef788a393f09c17858e10e42f61eb00f6a7ed96e#diff-9ddd9172e4a54cdd30d56e5f10d5facaR315

File: src/native_background.ts [hunk 6]
```diff
@@ -245,6 +246,17 @@ export async function temp(content: string, prefix: string) {
     return sendNativeMsg("temp", { content, prefix })
 }
 
+export async function winFirefoxRestart(profiledir: string) {
+    let current_version = await getNativeMessengerVersion()
+    let required_version = "0.1.6"
+
+    if (!await nativegate(required_version, false)) {
```
Comment (bovine3dom) [inline src/native_background.ts]: Native gate checks the version for you, so you're checking it twice here. It matters because we think it's quite slow - #512 - although I've never actually tested it. If you just set the second argument to true it will tell the user anyway. It won't say the name of the command but it will be right after they tell it to `restart` so it should be quite obvious.

File: src/native_background.ts (hunk 5 above)
Comment (bovine3dom) [reply src/native_background.ts]: Thanks! Can we delete this comment then?

Comment (gsbabil) [reply src/native_background.ts]: I've updated the comment [here][0] with further details. 

Note that Vimperator originally used to require escaping paths. So, there might be similar expectations from some users. Currently we are supporting profile directory paths _only_ and _exactly_ as displayed in `about:support`. 

I believe the rationale for this is that it will allows users to quickly copy-paste the profile directory path and set via `:set profiledir`, which I think also makes sense. We just need to communicate it clearly, and stay coherent in  the code-base.

Thanks.

[0]: https://github.com/gsbabil/tridactyl/commit/9ffc161b865d0ac6a5fe3712cc70cc74de3f2a98#diff-9ddd9172e4a54cdd30d56e5f10d5facaR298

File: src/native_background.ts (hunk 6 above)
Comment (gsbabil) [reply src/native_background.ts]: I originally _"stole"_ this snippet from [`getenv`][0], which had a similar pattern of redundantly calling `getNativeMessengerVersion()`. I patched both of them [here][1].

Thanks.

[0]: https://github.com/gsbabil/tridactyl/commit/9ffc161b865d0ac6a5fe3712cc70cc74de3f2a98#diff-9ddd9172e4a54cdd30d56e5f10d5facaL274
[1]: https://github.com/gsbabil/tridactyl/commit/9ffc161b865d0ac6a5fe3712cc70cc74de3f2a98#diff-9ddd9172e4a54cdd30d56e5f10d5facaR253

**Comment by bovine3dom** [comment]
Thanks, this looks good to go. I'll test it when I'm next by a Windows box - probably this evening - and merge it if all goes well.

File: native/native_main.py [hunk 7]
```diff
@@ -118,6 +130,197 @@ def sanitizeFilename(fn):
...
+
+    return is_valid
+
+
+def win_firefox_restart(message):
+    """Handle 'win_firefox_restart' message."""
+    reply = {}
+    profile_dir = message["profiledir"].strip()
+
+    ff_bin_name = "firefox.exe"
```
Comment (bovine3dom) [inline native/native_main.py]: We already have a setting for this used for `nativeopen`: `set browser`. It's for people using e.g. Nightly :)

File: native/native_main.py [hunk 8]
```diff
@@ -118,6 +130,197 @@ def sanitizeFilename(fn):
...
+            task_arg = "\"%s\"" % restart_ps1_path
+
+            open(restart_ps1_path, "w+").write(restart_ps1_content)
+
+            startupinfo = subprocess.STARTUPINFO()
+            startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+
+            subprocess.check_output(
+                ["powershell",
+                 "-NonInteractive",
```
Comment (bovine3dom) [inline native/native_main.py]: I added `-NoExit` here but all that seems to do is stop Firefox from exiting. The powershell window still flashes up unhelpfully.

Do you have any ideas for how I can debug this? I haven't touched profiledir: it's set to auto.

Comment (gsbabil) [reply native/native_main.py]: @bovine3dom

#### ## Look into `win_firefox_restart.ps1`

Could you please paste the content of `$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1` here? Check if you can run this script from a vanilla PowerShell prompt?

This is the script responsible to bring Firefox back. Note that this file is re-created every time `:restart` is invoked.

In PowerShell, the following command will copy the file into your clipboard. Please paste the content here to help debug further.

```Powershell
Get-Content "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1" | clip
```

#### ## Additional Debugging Steps

1. Keep Firefox (with the test profile) running, then in PowerShell run `&  "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"`

2. Keep Firefox closed, then in PowerShell run `&  "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"`

    - **[1]** Should put Powershell into a counting mode, waiting for the `parent.lock` file to be unlocked. Currently, this is set to `10 sec`, after that Powershell should show you an error message that Firefox needs to be restarted manually.

    - **[2]** Should immediately restart Firefox.

#### ## Background

Adding some more details below as reference. You might already know most of it. Hoping that someone else doing something similar might find it useful later on.

First of all, you'd find some verbose comments in `native_main.py::win_firefox_restart()` ([line 172][L172]) that both `Popen` family of functions exposed via `subprocess`, and `schtasks.exe` wasn't up to the task (at least for me on Windows). I still left them there, in case someone wants to take another stab at it using them later.

So, we end up doing the following _"orchestra"_ in JavaScript, Python and PowerShell:

1. User runs `:set profiledir [valid-profile-dir]`, followed by `:restart` in the JavaScript world.

   - JavaScript sends the following JSON object to the Python world:

```JSON
{
  "cmd":
        "win_firefox_restart",
  "profiledir":
        "C:\\Users\\<USERNAME>\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\x5gxqprf.dev-edition-default"
}

```

2. Upon receiving the `win_firefox_restart` command, followed by some sanity checks, Python (`native_main.py`) creates `win_firefox_restart.ps1` in `~/.tridactyl/` ([line 236][L236])

3. Python then runs a blocking `subprocess.check_output()` call ([line 282][L282]) invoking another set of PowerShell commands to achieve the following:

   - Schedule a Windows task (analogous to Unix [Cron][cron]) to be run _only once_ in `delay` amount of seconds (currently set to `1 sec`), which will invoke `win_firefox_restart.ps1` from a completely separate process group, so that Firefox can be revived back even after its "death".

If steps [1] and [2] are successful, we send `code = 0` to the JavaScript world, at which point JavaScript gracefully calls `qall()` killing itself.

```Python
  reply = {
      "code": 0,
      "content": "Restarting in %d seconds..." % delay_sec
  }
```

In the meantime, if `delay = 1 sec` has passed, Windows task scheduler invokes `win_firefox_restart.ps1` (_briefly showing the PowerShell screen - I think you are currently getting stuck here_), which basically means that Windows is attempting to restart Firefox via the `win_firefox_restart.ps1` script. If `parent.lock` is blocked, the PowerShell screen should display the counting at this point.

One key element here to be noted is that quitting of Firefox and restarting are being done _asynchronously_ by separate parent processes (with `parent.lock` being the safeguard), and the `win_firefox_restart.ps1` embeds the critical restart functionality on Windows. 

Let me know if it helps. I'm happy to share more, or join on IRC if needed.

Thanks.


[L172]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L172

[L236]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L236

[L282]: https://github.com/gsbabil/tridactyl/blob/ecc451ae323f3e10f5c96b7eab90c328276ccfe5/native/native_main.py#L282

[cron]: https://en.wikipedia.org/wiki/Cron

**Comment by gsbabil** [comment]
@bovine3dom 

I think I have found a corner-case when your `ExecutionPolicy` isn't set to `Bypass` system-wide. Could you please confirm the output of the following on your test machine:

```Powershell
> Get-ExecutionPolicy
```

Also, could you please run the following from an elevated PowerShell prompt, and try the `:restart` command. It'll help us narrow-down that `ExecutionPolicy` _is_ the actual problem.

```Powershell
> Set-ExecutionPolicy Bypass
```

Let me do some more test with `Set-ExecutionPolicy Restricted` and get back. Thanks.

File: native/native_main.py (hunk 7 above)
Comment (gsbabil) [reply native/native_main.py]: Hi @bovine3dom,

Correct me if I'm mistaken, but I don't think the _binary_ is named differently across different Firefox versions _e.g._ Stable, Beta, Developer, Nightly _etc._ all have `Firefox.exe`. The installation path is different across different versions, but the binary name remains the same AFAIK. `ff_bin_name = "firefox.exe"` above will use whichever `Firefox.exe` comes in `%PATH%` first, if the user has multiple versions of Firefox installed.

Are you saying that we should use the value found in `set browser` to restart Firefox with? For example, if the user set `:set browser c:\Program Files\Firefox Developer Edition\Firefox.exe`, `native_main.py` should receive that value when `:restart` is invoked, and use it during the actual restart? 

Please confirm, if the above is what you meant. I can send another patch to handle this. Thanks.

Comment (bovine3dom) [reply native/native_main.py]: Yes, I think so. It would be confusing if I `restart` Nightly and then get stable back, especially if I have set `firefox` to something else.

**Comment by bovine3dom** [comment]
> @bovine3dom
>
>I just submitted a patch here to fix :restart under restricted PowerShell. Please give it a go.
>
>Thanks.

Great, that's fixed the problem partially - `restart` now restarts Firefox.

However, when I run `restart` in a Firefox window that has been launched by `restart`, I get "#%firefox.exe not on path".

File: native/native_main.py [hunk 9]
```diff
@@ -118,6 +130,207 @@ def sanitizeFilename(fn):
...
+        #      "/TR",
+        #      "calc",
+        #      "/IT",
+        #      "/ST",
+        #      task_time],
+        #     shell=True)
+        # }}}
+
+        ff_bin_path = "\"%s\"" % shutil.which(ff_bin_name)
+        new_tab_url = "https://www.mozilla.org/en-US/firefox/"
```
Comment (bovine3dom) [inline native/native_main.py]: Why is this here? `restart` makes a new tab appear with this URL, which is a bit confusing. I expect `restart` just to be the same as closing Firefox and starting it again.

File: native/native_main.py [hunk 10]
```diff
@@ -118,6 +130,207 @@ def sanitizeFilename(fn):
...
+
+        if profile_dir == "auto":
+            ff_lock_path = ff_bin_path
+            ff_args = "\"%s\" \"%s\" \"%s\" \"%s\"" % \
+                ("-foreground",
+                 "-ProfileManager",
```
Comment (bovine3dom) [inline native/native_main.py]: Why do you have ProfileManager? Most people just use one profile, so seeing this appear will be a bit confusing. Is it because you have no way of knowing which profile is being used? I think I would rather not have it, as firefox will show it automatically if someone has not selected a default, which is the likely case if someone is using multiple profiles.

**Comment by bovine3dom** [comment]
Following your debugging steps anyway:

```powershell
Get-Content "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1" | clip
```
```powershell

$profileDir = "auto"
if ($profileDir -ne "auto") {
    $lockFilePath = "C:\Program Files\Mozilla Firefox\firefox.exe"
    $locked = $true
    $num_try = 10
} else {
    $locked = $false
}
while (($locked -eq $true) -and ($num_try -gt 0)) {
try {
    [IO.File]::OpenWrite($lockFilePath).close()
    $locked=$fals
} catch{
    $num_try-=1
    Write-Host "[+] Trial: $num_try [lock == true]"
    Start-Sleep -Seconds 1
}
}
if ($locked -eq $true) {
$errorMsg = "Restarting Firefox failed. Please restart manually."
Write-Host "$errorMsg"
# Add-Type -AssemblyName System.Windows.Forms
# [System.Windows.MessageBox]::Show(
#     $errorMsg,
#     "Tridactyl")
} else {
Write-Host "[+] Restarting Firefox ..."
& "C:\Program Files\Mozilla Firefox\firefox.exe" "-foreground" "-ProfileManager" "-new-tab" "https://www.mozilla.org/en-US/firefox/"
}

```

```powershell

PS C:\Windows\System32\WindowsPowerShell\v1.0> & "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"
```

```powershell
& : File C:\Users\olie\.tridactyl\win_firefox_restart.ps1 cannot be loaded because running scripts is disabled on this
system. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
At line:1 char:3
+ & "$env:USERPROFILE\.tridactyl\win_firefox_restart.ps1"
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : SecurityError: (:) [], PSSecurityException
    + FullyQualifiedErrorId : UnauthorizedAccess
```

**Comment by bovine3dom** [comment]
Before restarting:

```
! where firefox
C:\Program Files\Mozilla Firefox\firefox.exe 
```

**Comment by bovine3dom** [comment]
After restarting:

```
! where firefox
<blank line>
```

```
! dir
 Volume in drive C has no label. Volume Serial Number is 5427-04A0 Directory of C:\Users\olie\.tridactyl25/05/2018  20:40    <DIR>          .25/05/2018  20:40    <DIR>          ..15/05/2018  12:59    <DIR>          auto26/05/2018  13:54                65 native_main.bat26/05/2018  13:54            11,803 native_main.py26/05/2018  13:54               239 tridactyl.json26/05/2018  14:12               866 win_firefox_restart.ps1               4 File(s)         12,973 bytes               3 Dir(s)   9,869,664,256 bytes free 
```

```
! echo %PATH%
C:\Program Files\Oculus\Support\oculus-runtime;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\ProgramData\chocolatey\bin;C:\Program Files (x86)\MiKTeX 2.9\miktex\bin\;C:\Program Files (x86)\LyX 2.2\Perl\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Python27;C:\Python27\Scripts;C:\Program Files (x86)\Skype\Phone\;C:\Program Files (x86)\Mikero\DePboTools\bin;C:\Program Files (x86)\Pandoc\;C:\Program Files (x86)\Single-Drop-VIM\;C:\Program Files\Calibre2\;C:\Users\olie\AppData\Local\Julia-0.6.0\bin\;C:\Program Files (x86)\get_iplayer;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\nodejs\;C:\Program Files\Git\cmd;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Users\olie\AppData\Local\Programs\Python\Python36-32\Scripts\;C:\Users\olie\AppData\Local\Programs\Python\Python36-32\;C:\Users\olie\AppData\Local\Microsoft\WindowsApps;C:\ProgramData\chocolatey\lib\msys2;C:\tools\msys64;C:\Program Files (x86)\Oni;C:\Users\olie\AppData\Roaming\npm 
```

File: native/native_main.py (hunk 10 above)
Comment (gsbabil) [reply native/native_main.py]: [Done][0].

[0]: https://github.com/gsbabil/tridactyl/commit/b6b71c44678cee08b3e6a7b8472a15c82b9e0ebf

File: native/native_main.py (hunk 9 above)
Comment (gsbabil) [reply native/native_main.py]: [Removed][0].

[0]: https://github.com/gsbabil/tridactyl/commit/0584d42d62a56f89f53d40bea18de77ba3c36791

**Comment by gsbabil** [comment]
```
C:\Users\olie\.tridactyl\win_firefox_restart.ps1 cannot be loaded
because running scripts is disabled on this system.
```

I think it was the `ExecutionPolicy` set on your system, and should be fixed now with [this patch][0].

[0]: https://github.com/gsbabil/tridactyl/commit/93ca6c50a694bbaf6d093c37398856d10010ee9c

**Comment by gsbabil** [comment]
@bovine3dom 

> However, when I run restart in a Firefox window that has been launched by restart, I get #firefox.exe is not found on %PATH%..

That's weird. The check that leads to this is basically just the following. It looks like something is overwriting your `%PATH%`.

```Python
if shutil.which("firefox.exe")
```

I can't immediately reproduce this on my system. Could you please confirm how you are collecting the `%PATH%` values below. For example,  is it like you `:restart` using Tridactyl, then open a PowerShell or `cmd.exe` prompt, and run `where firefox.exe`?

```
! where firefox
C:\Program Files\Mozilla Firefox\firefox.exe 

[...]

! where firefox
<blank line>
```

Given the data above, after `:restart` you `%PATH%` somehow becomes the following, which does not seem to have Firefox anymore!

```
C:\Program Files\Oculus\Support\oculus-runtime
C:\ProgramData\Oracle\Java\javapath
C:\Windows\system32
C:\Windows
C:\Windows\System32\Wbem
C:\Windows\System32\WindowsPowerShell\v1.0\
C:\ProgramData\chocolatey\bin
C:\Program Files (x86)\MiKTeX 2.9\miktex\bin\
C:\Program Files (x86)\LyX 2.2\Perl\bin
C:\WINDOWS\system32
C:\WINDOWS
C:\WINDOWS\System32\Wbem
C:\WINDOWS\System32\WindowsPowerShell\v1.0\
C:\Python27
C:\Python27\Scripts
C:\Program Files (x86)\Skype\Phone\
C:\Program Files (x86)\Mikero\DePboTools\bin
C:\Program Files (x86)\Pandoc\
C:\Program Files (x86)\Single-Drop-VIM\
C:\Program Files\Calibre2\
C:\Users\olie\AppData\Local\Julia-0.6.0\bin\
C:\Program Files (x86)\get_iplayer
C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common
C:\Program Files\nodejs\
C:\Program Files\Git\cmd
C:\WINDOWS\system32
C:\WINDOWS
C:\WINDOWS\System32\Wbem
C:\WINDOWS\System32\WindowsPowerShell\v1.0\
C:\WINDOWS\System32\OpenSSH\
C:\Users\olie\AppData\Local\Programs\Python\Python36-32\Scripts\
C:\Users\olie\AppData\Local\Programs\Python\Python36-32\
C:\Users\olie\AppData\Local\Microsoft\WindowsApps
C:\ProgramData\chocolatey\lib\msys2
C:\tools\msys64
C:\Program Files (x86)\Oni
C:\Users\olie\AppData\Roaming\npm
```

**Comment by bovine3dom** [comment]
> I can't immediately reproduce this on my system. Could you please confirm how you are collecting the %PATH% values below. For example, is it like you :restart using Tridactyl, then open a PowerShell or cmd.exe prompt, and run where firefox.exe? Do you close the PowerShell/cmd.exe window between restart?
> 
> Please share some more details. Let me see if I can reproduce this to debug further.

This is in Tridactyl: `!` is an alias for `exclaim` which executes things in the system shell via the native messenger.

I start Firefox normally (as in, I click the shortcut Mozilla placed in the start menu), run the commands, `restart`, and then run the commands again.

Firefox is not normally on my path, but if I do `! start cmd.exe` I get a shell which does have Firefox on its path.


Feel free to come on IRC/Matrix (the bridge between the two very occasionally is unreliable, so just try sending your message again if you get no response).

**Comment by gsbabil** [comment]
@bovine3dom 

> I start Firefox normally (as in, I click the shortcut Mozilla placed in the start menu), run the commands, restart, and then run the commands again.

> Firefox is not normally on my path (i.e, if I start cmd.exe from the start menu), but if I do ! start cmd.exe I get a shell which does have Firefox on its path, so long as it isn't a restarted process.

I was able to debug this to the [following][image], and just submitted 47d247a that doesn't require Firefox to be on path anymore (although, I'd advise for it, particularly if the user wants to do `Win-R > firefox -P` type of activities :-). 47d247a  also respects Tridactyl's `set browser [browser-cmd]` setting, and avoids using hard-coded `firefox.exe`.

Basically, clicking the Windows Firefox short-cut link conveniently puts the Firefox directory on path (analogous to `cd /custom/firefox-dir && ./firefox.bin`), which results into the "positive" outcome of the first `! where firefox` when Firefox is started by clicking the short-cut `LNK` file. 

In 9d5ce40, I was not setting `CWD` and directly invoking the Firefox binary, which explains the "negative" outcome of the second `! where firefox` after restart.

Thank you.

image

[image]: https://user-images.githubusercontent.com/1061108/40587191-29731926-620f-11e8-8b1f-367741d69ef2.png
